### PR TITLE
Default to enforced auth in tests

### PR DESCRIPTION
**Pull request by leighmcculloch**
### What

Default to enforced auth in tests.

Also renamed 'record' to 'mock' so clarity that when in use, these functions are mocking auth.

Also as a side-effect the auths() fn that returns a list of seen auths during an invocation now works for enforced auth too.

### Example

```rust
#[contractimpl]
impl Contract {
    pub fn doit(a: Address) -> u64 {
        a.require_auth();
        2
    }
}
```
```rust
#[test]
fn test_with_mock_auth() {
    let e = Env::default();
    
    e.mock_all_auths()

    let contract_id = e.register_contract(None, Contract);
    let client = ContractClient::new(&e, &contract_id);

    let a = Address::random(&e);

    let r = client.doit(&a);

    assert_eq!(r, 2);
}
```

### Why

The test framework doesn't allow the developer to use auth like it will behave in production. As much as possible a principle of the test framework is that it behaves like production. While we have and are improving functions for mocking auth such that testing with auth is convenient, the test framework should also be able to behave like production.

Dependent on https://github.com/stellar/rs-soroban-env/pull/787

Close https://github.com/stellar/rs-soroban-sdk/issues/917

### CODE REVIEW

File: soroban-sdk/src/env.rs [hunk 1]
```diff
@@ -750,20 +811,10 @@ impl Env {
...
-    /// to coincide with the actual top-level contract invocation. For example,
-    /// if contract A doesn't use `require_auth` and then it calls contract B
-    /// that uses `require_auth`, then `verify_top_authorization` will return
-    /// `true` when verifying the contract B's `require_auth` call, but it will
-    /// return `false` if contract A makes a `require_auth` call.
-    ///
-    /// It is possible for a single address to be present multiple times in the
-    /// output, as long as there are multiple disjoint call trees for that
-    /// address.
+    /// [`require_auth`] calls. Repeated calls to [`require_auth`] in the same
```
Comment (dmkozh) [inline soroban-sdk/src/env.rs:814]: > Repeated calls to [`require_auth`] in the same tree of contract invocations will appear only once in the vector

This is not correct. First, trees are relative to the address, so for multiple addresses multiple entries will be present for the same contract invocation. Second, it's actually possible to call `reqiure_auth` for the same address multiple times in the same call - that's fine as long as the arguments are different.

File: soroban-sdk/src/env.rs [hunk 2]
```diff
@@ -737,8 +736,70 @@ impl Env {
...
+    ///     let client = HelloContractClient::new(&env, &contract_id);
+    ///     let addr = Address::random(&env);
+    ///     client.hello(&addr);
+    /// }
+    /// ```
+    pub fn mock_all_auths(&self) {
+        self.env_impl.switch_to_recording_auth();
+    }
+
+    /// Returns a list of authorizations that were seen during the last contract
```
Comment (dmkozh) [inline soroban-sdk/src/env.rs:801]: These are still 'top' authorizations, but the comment doesn't really refer to that. While the previous definition probably had some room to improve the clarity, it contained information about what the 'top' means. That's important as e.g. internal token transfer calls won't be visible here and this is probably the most frequent scenario for the auth trees.

File: soroban-sdk/src/env.rs (hunk 1 above)
Comment (leighmcculloch) [reply soroban-sdk/src/env.rs]: Fixed in d66eb844 and 53266a03. Thanks!

File: soroban-sdk/src/env.rs (hunk 2 above)
Comment (leighmcculloch) [reply soroban-sdk/src/env.rs:801]: Thanks for identifying this gap in the docs. I'd like to address the gap by removing top-level auths from the SDK for this function, to simplify the concepts a developer has to interact with. Opened https://github.com/stellar/rs-soroban-env/pull/787 and updated this pull request to use it instead. This PR is blocked on that env change.

**Pull request merged**